Microsoft.Toolkit.Mvvm/ComponentModel/ObservableRecipient.cs

```diff
@@ -0,0 +1,234 @@
+        /// <summary>
+        /// Compares the current and new values for a given property. If the value has changed,
+        /// raises the <see cref="ObservableObject.PropertyChanging"/> event, updates the property with
+        /// the new value, then raises the <see cref="ObservableObject.PropertyChanged"/> event.
+        /// </summary>
+        /// <typeparam name="T">The type of the property that changed.</typeparam>
+        /// <param name="field">The field storing the property's value.</param>
+        /// <param name="newValue">The property's value after the change occurred.</param>
+        /// <param name="broadcast">If <see langword="true"/>, <see cref="Broadcast{T}"/> will also be invoked.</param>
+        /// <param name="propertyName">(optional) The name of the property that changed.</param>
+        /// <returns><see langword="true"/> if the property was changed, <see langword="false"/> otherwise.</returns>
+        /// <remarks>
+        /// This method is just like <see cref="ObservableObject.Set{T}(ref T,T,string)"/>, just with the addition
+        /// of the <paramref name="broadcast"/> parameter. As such, following the behavior of the base method,
+        /// the <see cref="ObservableObject.PropertyChanging"/> and <see cref="ObservableObject.PropertyChanged"/> events
+        /// are not raised if the current and new value for the target property are the same.
+        /// </remarks>
+        protected bool Set<T>(ref T field, T newValue, bool broadcast, [CallerMemberName] string propertyName = null!)
+        {
+            return Set(ref field, newValue, EqualityComparer<T>.Default, broadcast, propertyName);
+        }
+
+        /// <summary>
+        /// Compares the current and new values for a given property. If the value has changed,
+        /// raises the <see cref="ObservableObject.PropertyChanging"/> event, updates the property with
+        /// the new value, then raises the <see cref="ObservableObject.PropertyChanged"/> event.
+        /// See additional notes about this overload in <see cref="Set{T}(ref T,T,bool,string)"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of the property that changed.</typeparam>
+        /// <param name="field">The field storing the property's value.</param>
+        /// <param name="newValue">The property's value after the change occurred.</param>
+        /// <param name="comparer">The <see cref="IEqualityComparer{T}"/> instance to use to compare the input values.</param>
+        /// <param name="broadcast">If <see langword="true"/>, <see cref="Broadcast{T}"/> will also be invoked.</param>
+        /// <param name="propertyName">(optional) The name of the property that changed.</param>
+        /// <returns><see langword="true"/> if the property was changed, <see langword="false"/> otherwise.</returns>
+        protected bool Set<T>(ref T field, T newValue, IEqualityComparer<T> comparer, bool broadcast, [CallerMemberName] string propertyName = null!)
```

Comment:
I know you like to have short names but as `set` is also a keyword, reusing it as a method name can cause unnecessary confusion and complexity in languages other than C#.
I know that C# is much more popular than Visual Basic but VB still accounts for at least 10% of UWP development and much more for WPF.
Naming this method `SetProperty`  (as some other MVVM frameworks do)  would:
- Improve clarification of what is being "set"
- Improve _inclusivity_ of experience for developers who don't use C#

(This also applies to ObservableObject too)

Reply:
Ah, yeah this is something where I'll have to disagree 😅

I did give this some thought in the past as well and I do believe `Set` is the right name, for a number of reasons:

- `SetProperty` doesn't actually represent what all overloads are doing, as some are more accurately just setting a field and then raising a property changed event. For instance, having a "SetProperty" method taking a ref to a field as first argument would look very weird to me, and that's also the overload that we'd generally assume would get used the most.
- You can still use `Set` on VB.NET, just need `MyBase.Set` (maybe even just `Me.Set`?). I don't think trading verbosity on C# for VB.NET is a good idea given the low usage and the fact that it's basically deprecated. If one of the two languages should have a less convenient version between the two, it should be VB.NET, not C#.
- This is one of the methods that will be used more often (I mean considering the overloads too), and having a less verbose and more compact name really makes a difference over time.
- "Set" is perfectly consistent with the actual set keyword being used in property setters, as in: `set => Set(...)`.
- "Set" is also consistent with the name originally used in `MvvmLight`. I wouldn't want people migrating from there to get a more verbose codebase as a result.
- Also, for consistency we'd then also need to use `SetPropertyAndNotifyOnCompletion`, which is both just incredible verbose at that point, and also has again the issue that actually sets a field and not a property. Yes the field backs a property, but you're still effectively using a method called "SetProperty" to really just set a field instead.

Hope this makes sense! 😊

Reply:
> Ah, yeah this is something where I'll have to disagree 😅
> 
> I did give this some thought in the past as well and I do believe `Set` is the right name, for a number of reasons:
> 
> * `SetProperty` doesn't actually represent what all overloads are doing, as some are more accurately just setting a field and then raising a property changed event. For instance, having a "SetProperty" method taking a ref to a field as first argument would look very weird to me, and that's also the overload that we'd generally assume would get used the most.
> * You can still use `Set` on VB.NET, just need `MyBase.Set` (maybe even just `Me.Set`?). I don't think trading verbosity on C# for VB.NET is a good idea given the low usage and the fact that it's basically deprecated. If one of the two languages should have a less convenient version between the two, it should be VB.NET, not C#.
> * This is one of the methods that will be used more often (I mean considering the overloads too), and having a less verbose and more compact name really makes a difference over time.
> * "Set" is perfectly consistent with the actual set keyword being used in property setters, as in: `set => Set(...)`.
> * "Set" is also consistent with the name originally used in `MvvmLight`. I wouldn't want people migrating from there to get a more verbose codebase as a result.
> * Also, for consistency we'd then also need to use `SetPropertyAndNotifyOnCompletion`, which is both just incredible verbose at that point, and also has again the issue that actually sets a field and not a property. Yes the field backs a property, but you're still effectively using a method called "SetProperty" to really just set a field instead.
> 
> Hope this makes sense! 😊

It's your project and so your choice, but I'll leave this discussion with a few clarifications.

- It probably comes down to preference but there seems to be a broad divide on this generally. Prism, Calcium, & MvvmCross use `SetProperty`. MVVMLight, Caliburn.Micro, & MVVM Basic (copied from MVVM Light) use `Set`. By adopting `SetProperty` here, it would leave a broad consensus among frameworks with Caliburn.Micro as the only dissenter. Helping to standardize the term used everywhere can only be a positive thing.

- VB.NET is **NOT** "basically deprecated" and this kind of FUD is unhelpful to repeat. Using that as an argument for not considering it as a valid language show a lack of understanding of the language you're making decisions about.

- As a principle, rejecting a change that will increase consistency and avoid potential problems for a minority of a population because they will mean a small change for the majority of a population should only be done with great care. If the person making the decision is part of the majority, consultation with members of the minority is vital.

- Claiming a desire for consistency with MVVM Light now, when it wasn't a concern previously (you even claimed the start of this new library was a good time to introduce breaking changes if appropriate) seems inconsistent. Are you now looking at this as a natural successor to MVVM light and expecting people to switch? If so, documentation for this is essential and tracking potential differences is something that should have been done from the start.

Reply:
Thanks @mrlacey for the breakdown here. It's good to know about which libraries use which terms. I do like the idea of aligning with other large libraries out there, though part of this whole project did come from the idea of having something to replace MVVMLight.

@jamesmcroft thoughts on this terminology? Do you think there's enough other things from someone migrating from MVVMLight that another name change like this one would be a burden? If not, then I'd say let's go with `SetProperty` as Caliburn.Micro is being deprecated, so then all the MVVM Libraries would be aligned and help any folks wanting to migrate between them, as this is really the biggest pattern used in MVVM, eh?

As for VB.NET it's never been a focus for the Toolkit, but I do appreciate all the work you've done @mrlacey to help us provide samples and let us know where things are broken. @Sergio0694 I know they've said they're not adding new features to the language, but they haven't said it's not supported or given an indication for folks to not develop for it. Let's make sure we're all respectful of all the ways in which our technologies can be used. There's so many out there and so many creative ways to use them together, it's hard to think of all these different cases sometimes where our code will be executed.

Reply:
> they haven't said it's not supported or given an indication for folks to not develop for it

They've done the opposite. They have committed to maintaining support and tooling so people can keep using it. (Although not in every project type or framework)

Reply:
@michael-hawker I don't think it's a problem to rename this but I think the naming has to make sense. I've only really ever used MvvmLight for my projects so it's going to be a hard sell to change it to `SetProperty` given that it doesn't necessarily always set a property as pointed out by @Sergio0694. 

Although trying to be consistent with other frameworks terminology helps, it doesn't always necessarily mean it's right. This is where I lean towards leaving this as `Set` but I'm not precious on this. 

As long as the naming is there, it will be easy to write migration documentation which aligns to the APIs.

Reply:
It does need to be taken into consideration also that while this project is a spawn of wanting to maintain the MvvmLight codebase, it may not just be those from MvvmLight who wish to migrate to the WCT framework. Just a little extra food for thought. 

Ultimately, I'll let the general consensus decide here and we can work around it in regards to any migration documentation. 

On that note, I've been holding off doing too much until things have been finalized with the APIs. When do we think may be a good time to start looking into the migration docs for those who are currently using MvvmLight and may wish to move over? Happy to start ASAP/when needed 👍

Reply:
Just taken a look at the APIs for Prism and MvvmCross to understand more around the usages of `SetProperty`. I still don't think it's naming is right for the action it performs. That being said, aside from MvvmLight, these two, in particular, are the most common go-to MVVM frameworks that I know are floated around with app development. 

As previously stated, I'm easy either way on this one. As long as we can nail the migration docs for those using the MvvmLight framework, it shouldn't be an issue at all 😄

Reply:
Thanks @jamesmcroft for the input. Let's ask around the discord community a bit for the naming to see if anything strikes a balance. The argument for consistency is at least at the forefront of my mind for this. I think in terms of verbosity with templates and copy/paste I'm not as worried about that side of things.

I think once this PR is merged, it'd be a safe time to start working on migration docs. At least that'll start getting us some more input from that community of folks as we ship a preview ahead of final release, and we can work from there on any changes all-up with docs/samples/library/migration guide?

Reply:
Resolving this thread as we agreed to move to `SetProperty` here, thanks @mrlacey!